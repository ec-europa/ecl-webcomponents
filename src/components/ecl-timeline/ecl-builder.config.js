const path = require('path');
const pkg = require('./package.json');

const nodeModules = path.resolve(__dirname, '../../../node_modules');
const outputFolder = './build';

// SCSS includePaths
const includePaths = [nodeModules];

const banner = `${pkg.name} - ${pkg.version} Built on ${new Date().toISOString()}`;

module.exports = {
  scripts: [
    {
      entry: path.resolve(nodeModules, '@ecl/vanilla-component-timeline/timeline.js'),
      dest: path.resolve(outputFolder, 'scripts/ecl-timeline-vanilla.js'),
      options: {
        banner,
        moduleName: 'ECL',
        includePaths,
        sourceMap: false,
      },
    },
  ],
  styles: [
    {
      entry: path.resolve(__dirname, 'src/styles/ecl-timeline-ec.scss'),
      dest: path.resolve(outputFolder, 'styles/ecl-timeline-ec.css'),
      options: {
        banner,
        includePaths,
        sourceMap: false,
      },
    },
    {
      entry: path.resolve(__dirname, 'src/styles/ecl-timeline-eu.scss'),
      dest: path.resolve(outputFolder, 'styles/ecl-timeline-eu.css'),
      options: {
        banner,
        includePaths,
        sourceMap: false,
      },
    },
  ],
};
