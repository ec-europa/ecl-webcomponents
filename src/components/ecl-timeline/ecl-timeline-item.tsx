import { Component, h, Prop} from '@stencil/core';

@Component({
  tag: 'ecl-timeline-item',
  shadow: false,
  scoped: false,
})
export class EclTimelineItem {
  @Prop() theme: string = 'ec';
  @Prop() styleClass: string;
  @Prop() label: string;
  @Prop() type: string;
  @Prop() itemTitle: string;

  getClass(): string {
    const styleClasses = [
      `ecl-timeline__item`,
      this.styleClass
    ];

    if (this.type === 'toggle') {
      styleClasses.push('ecl-timeline__item--toggle');
    }

    return styleClasses.join(' ');
  }

  render() {
    return (
      <li class={this.getClass()}>
      { this.type !== 'toggle' ?
        <div class={`ecl-timeline__tooltip`}>
          <div class={`ecl-timeline__tooltip-arrow`}></div>
        { this.label ?
          <div class={`ecl-timeline__label`}>
            {this.label}
          </div> : ''
        }
        { this.itemTitle ?   
          <div class={`ecl-timeline__title`}>
            {this.itemTitle}
          </div> : ''
        }
          <div class={`ecl-timeline__content`}>
            <slot></slot>
          </div>
        </div> : '' 
      }
      { this.type === 'toggle' ?
        <ecl-button
          type="secondary"
          data-ecl-timeline-button
        >
          <ecl-icon
            styleClass="ecl-button__icon--after"
            icon="corner-arrow"
            size="xs"
            transform="rotate-180"
            slot="icon-after"
          ></ecl-icon>
          <slot></slot>
      </ecl-button> : ''
      }
      </li>
    );
  }
}