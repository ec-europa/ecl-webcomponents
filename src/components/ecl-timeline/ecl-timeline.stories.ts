import { randomizedLink } from "../../utils/randomizedLink";

export default {
  title: 'Components/timeline',
};

const Template = (args) => 
  `<ecl-timeline
    theme="${args.theme}"
    ecl-script
   >
    <ecl-timeline-item
      label="13 September 2017"
      item-title="Item title"
      itemId="0"
      theme="${args.theme}"
    >
      <ecl-link theme="${args.theme}" path="${randomizedLink('/example.html')}">President Juncker's State of the Union speech</ecl-link>
    </ecl-timeline-item>
    <ecl-timeline-item
      item-id="1"
      label="28-29 September 2017"
      theme="${args.theme}"
    >
      <ecl-link theme="${args.theme}" path="${randomizedLink('/example.html')}">Informal Digital Summit, Tallinn</ecl-link>
    </ecl-timeline-item>
    <ecl-timeline-item
      item-id="2"
      label="14 November 2017"
      theme="${args.theme}"
    >
      <ecl-link path="${randomizedLink('/example.html')}">Strengthening European identity through education and culture: European Commission's contribution to the Leaders' meeting, Gothenburg, Sweden</ecl-link>
    </ecl-timeline-item>
    <ecl-timeline-item
      item-id="3"
      label="17 November 2017"
      theme="${args.theme}"
    >
      <ecl-link theme="${args.theme}" path="${randomizedLink('/example.html')}">Social Summit in Gothenburg, Sweden</ecl-link>
    </ecl-timeline-item>
    <ecl-timeline-item
      item-id="4"
      label="6 December 2017"
      theme="${args.theme}"
    >
      <ecl-link theme="${args.theme}" path="${randomizedLink('/example.html')}">Economic and Monetary Union package of proposals</ecl-link>
    </ecl-timeline-item>
    <ecl-timeline-item 
      type="toggle"
      theme="${args.theme}"
    >
      View all
    </ecl-timeline-item>
    <ecl-timeline-item
      item-id="5"
      label="14-15 December 2017"
      style-class="ecl-timeline__item--collapsed"
      theme="${args.theme}"
    >
      <ecl-link path="${randomizedLink('/example.html')}">EU Leaders' meeting on migration, Brussels</ecl-link>
    </ecl-timeline-item>
    <ecl-timeline-item
      item-id="6"
      label="15 December 2017"
      theme="${args.theme}"
    >
      <ecl-link theme="${args.theme}" path="${randomizedLink('/example.html')}">Euro Summit</ecl-link>
    </ecl-timeline-item>
    <ecl-timeline-item
      item-id="7"
      label="6 February 2018"
      theme="${args.theme}"
    >
      <ecl-link theme="${args.theme}" path="${randomizedLink('/example.html')}">EU-Western Balkans Strategy</ecl-link>
    </ecl-timeline-item>
    <ecl-timeline-item
      item-id="8"
      label="14 February 2018"
      theme="${args.theme}"
    >
      <ecl-link path="${randomizedLink('/example.html')}">Multiannual Financial Framework</ecl-link> and <ecl-link path="${randomizedLink('/example.html')}">institutional issues</ecl-link> - enhancing efficiency at the helm of the European Union
    </ecl-timeline-item>
    <ecl-timeline-item
      item-id="9"
      label="23 February 2018"
      theme="${args.theme}"
    >
      <ecl-link path="${randomizedLink('/example.html')}">Informal European Council</ecl-link>
    </ecl-timeline-item>
    <ecl-timeline-item
      item-id="10"
      label="22-23 March 2018"
      theme="${args.theme}"
    >
      <ecl-link path="${randomizedLink('/example.html')}">European Council</ecl-link>
    </ecl-timeline-item>
    <ecl-timeline-item
      item-id="11"
      label="2 May 2018"
      theme="${args.theme}"
    >
      <ecl-link path="${randomizedLink('/example.html')}">Commission proposal: Long-term budget post-2020</ecl-link>
    </ecl-timeline-item>
    <ecl-timeline-item
      item-id="12"
      label="15 May 2018"
      theme="${args.theme}"
    >
      <ecl-link path="${randomizedLink('/example.html')}">European Commissionâ€™s contribution to the Informal Leaders' meeting</ecl-link>, Sofia, Bulgaria (16 May 2018)
    </ecl-timeline-item>
    <ecl-timeline-item
      item-id="13"
      label="17 May 2018"
      theme="${args.theme}"
    >
      <ecl-link theme="${args.theme}" path="${randomizedLink('/example.html')}">EU-Western Balkans Summit</ecl-link> (Sofia, Bulgaria)
    </ecl-timeline-item>
  </ecl-timeline>`;

export const Timeline = Template.bind({});
Timeline.storyName = 'default';