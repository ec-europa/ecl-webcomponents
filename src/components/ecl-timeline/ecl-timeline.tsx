import { Component, h, Prop, Element, State } from '@stencil/core';
import getAssetPath from "../../utils/assetPath";
declare const ECL: any;

@Component({
  tag: 'ecl-timeline',
  styleUrls: {
    ec: './build/styles/ecl-timeline-ec.css',
    eu: './build/styles/ecl-timeline-eu.css',
  },
  shadow: false,
  scoped: true,
  assetsDirs: ['build'],
})
export class EclTimeline {
  @Element() el: HTMLElement;
  @Prop() theme: string = 'ec';
  @Prop() styleClass: string;
  @Prop() eclScript: boolean = false;
  @State() toBeToggled: boolean = false;

  getClass(): string {
    const styleClasses = [
      `ecl-timeline`,
      this.styleClass
    ];

    return styleClasses.join(' ');
  }

  componentDidRender() {
    const items = this.el.querySelectorAll('.ecl-timeline__item');
    this.el.querySelector('.ecl-timeline').innerHTML = '';
    this.el.querySelector('.ecl-timeline').append(...items);

    if (this.eclScript) {
      // Load the ECL vanilla js if not already present.
      const src = getAssetPath('./build/scripts/ecl-timeline-vanilla.js');
      if (document.querySelector(`script[src="${src}"]`)) {
        document.querySelector(`script[src="${src}"]`).remove();
      }
      const script = document.createElement('script');
      script.src = src;
      script.onload = () => {
        const timeline = new ECL.Timeline(this.el.firstElementChild);
        timeline.init();
      };

      document.body.appendChild(script);
    }

    const toggle = this.el.querySelector('.ecl-timeline__item--toggle');

    if (toggle) {
      toggle.addEventListener('click', () => {
        this.toBeToggled = true;
      });
    }
  }

  render() {
    return (
      <ol 
        class={this.getClass()}
        data-ecl-timeline
      >
        <slot></slot>
      </ol>
    );
  }
}