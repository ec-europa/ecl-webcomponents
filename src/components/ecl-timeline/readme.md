# ecl-timeline



<!-- Auto Generated Below -->


## Properties

| Property     | Attribute     | Description | Type     | Default     |
| ------------ | ------------- | ----------- | -------- | ----------- |
| `itemTitle`  | `item-title`  |             | `string` | `undefined` |
| `label`      | `label`       |             | `string` | `undefined` |
| `styleClass` | `style-class` |             | `string` | `undefined` |
| `theme`      | `theme`       |             | `string` | `'ec'`      |
| `type`       | `type`        |             | `string` | `undefined` |


## Dependencies

### Depends on

- [ecl-button](../ecl-button)
- [ecl-icon](../ecl-icon)

### Graph
```mermaid
graph TD;
  ecl-timeline-item --> ecl-button
  ecl-timeline-item --> ecl-icon
  ecl-button --> ecl-indicator
  style ecl-timeline-item fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
