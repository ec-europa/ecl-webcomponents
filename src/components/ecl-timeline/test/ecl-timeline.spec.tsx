import { newSpecPage } from '@stencil/core/testing';
import { EclTimeline } from '../ecl-timeline';
import { EclTimelineItem } from '../ecl-timeline-item';

describe('ecl-timeline', () => {
  it('renders', async () => {
    const page = await newSpecPage({
      components: [EclTimeline, EclTimelineItem],
      html: `ecl-timeline
        theme="ec"
        ecl-script
       >
        <ecl-timeline-item
          label="13 September 2017"
          item-title="Item title"
          itemId="0"
        >
          <a href="/example.html" class="ecl-link">President Juncker's State of the Union speech</a>
        </ecl-timeline-item>
        <ecl-timeline-item
          item-id="1"
          label="28-29 September 2017"
        >
          <a href="/example.html" class="ecl-link">Informal Digital Summit, Tallinn</a>
        </ecl-timeline-item>
        <ecl-timeline-item
          item-id="2"
          label="14 November 2017"
        >
          <a href="/example.html" class="ecl-link">Strengthening European identity through education and culture: European Commission's contribution to the Leaders' meeting, Gothenburg, Sweden</a>
        </ecl-timeline-item>
        <ecl-timeline-item
          item-id="3"
          label="17 November 2017"
        >
          <a href="/example.html" class="ecl-link">Social Summit in Gothenburg, Sweden</a>
        </ecl-timeline-item>
        <ecl-timeline-item
          item-id="4"
          label="6 December 2017"
        >
          <a href="/example.html" class="ecl-link">Economic and Monetary Union package of proposals</a>
        </ecl-timeline-item>
        <ecl-timeline-item type="toggle">
          Show/Hide hidden items
        </ecl-timeline-item>
        <ecl-timeline-item
          item-id="5"
          label="14-15 December 2017"
          style-class="ecl-timeline__item--collapsed"
        >
          <a href="/example.html" class="ecl-link">EU Leaders' meeting on migration, Brussels</a>
        </ecl-timeline-item>
        <ecl-timeline-item
          item-id="6"
          label="15 December 2017"
          style-class="ecl-timeline__item--collapsed"
        >
          <a href="/example.html" class="ecl-link">Euro Summit</a>
        </ecl-timeline-item>
        <ecl-timeline-item
          item-id="7"
          label="6 February 2018"
          style-class="ecl-timeline__item--collapsed"
        >
          <a href="/example.html" class="ecl-link">EU-Western Balkans Strategy</a>
        </ecl-timeline-item>
        <ecl-timeline-item
          item-id="8"
          label="14 February 2018"
          style-class="ecl-timeline__item--collapsed"
        >
          <a href="/example.html" class="ecl-link">Multiannual Financial Framework</a> and <a href="/example.html" class="ecl-link">institutional issues</a> - enhancing efficiency at the helm of the European Union
        </ecl-timeline-item>
        <ecl-timeline-item
          item-id="9"
          label="23 February 2018"
          style-class="ecl-timeline__item--collapsed"
        >
          <a href="/example.html" class="ecl-link">Informal European Council</a>
        </ecl-timeline-item>
        <ecl-timeline-item
          item-id="10"
          label="22-23 March 2018"
          style-class="ecl-timeline__item--collapsed"
        >
          <a href="/example.html" class="ecl-link">European Council</a>
        </ecl-timeline-item>
        <ecl-timeline-item
          item-id="11"
          label="2 May 2018"
          style-class="ecl-timeline__item--collapsed"
        >
          <a href="/example.html" class="ecl-link">Commission proposal: Long-term budget post-2020</a>
        </ecl-timeline-item>
        <ecl-timeline-item
          item-id="12"
          label="15 May 2018"
          style-class="ecl-timeline__item--collapsed"
        >
          <a href="/example.html" class="ecl-link">European Commissionâ€™s contribution to the Informal Leaders' meeting</a>, Sofia, Bulgaria (16 May 2018)
        </ecl-timeline-item>
        <ecl-timeline-item
          item-id="13"
          label="17 May 2018"
          style-class="ecl-timeline__item--collapsed"
        >
          <a href="/component-library/example" class="ecl-link">EU-Western Balkans Summit</a> (Sofia, Bulgaria)
        </ecl-timeline-item>
      </ecl-timeline>`,
    });

    expect(page.root).toMatchSnapshot();
  });
});
