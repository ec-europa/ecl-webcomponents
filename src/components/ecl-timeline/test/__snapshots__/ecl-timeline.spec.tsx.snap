// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ecl-timeline renders 1`] = `
<ecl-timeline-item item-title="Item title" itemid="0" label="13 September 2017">
  <!---->
  <li class="ecl-timeline__item sc-ecl-timeline-ec">
    <div class="ecl-timeline__tooltip sc-ecl-timeline-ec">
      <div class="ecl-timeline__tooltip-arrow sc-ecl-timeline-ec"></div>
      <div class="ecl-timeline__label sc-ecl-timeline-ec">
        13 September 2017
      </div>
      <div class="ecl-timeline__title sc-ecl-timeline-ec">
        Item title
      </div>
      <div class="ecl-timeline__content sc-ecl-timeline-ec">
        <a class="ecl-link" href="/example.html">
          President Juncker's State of the Union speech
        </a>
      </div>
    </div>
  </li>
</ecl-timeline-item>
`;
